##### 1. Write a Python program to check that a string contains only lowercase and uppercase characters.

In [12]:
import re

In [ ]:
pattern = r"^[a-zA-Z]+$"
string = "sdfsdfSDFSDF"
if match := re.search(pattern, string):
    print(match)
else:
    print("pattern not found")

<re.Match object; span=(0, 12), match='sdfsdfSDFSDF'>


##### 2. Write a Python program that matches a word at the beginning of a string.

In [ ]:
pattern = r"^[a-zA-Z]+\b"
string = "sdf sdf sdf sdf sdf"
if match := re.search(pattern, string):
    print(match)
else:
    print("pattern not found")

<re.Match object; span=(0, 3), match='sdf'>


##### 3. Write a Python program that matches a word at the end of string

In [ ]:
pattern = r"\b[a-zA-Z]+$"
string = "sdf sdf sdf sdf lastword"
if match := re.search(pattern, string):
    print(match)
else:
    print("pattern not found")

<re.Match object; span=(16, 24), match='lastword'>


##### 4. Write a Python program that matches a word containing 'z', not at the start or end of the word. (Matches Brazil but not Zelda)

In [ ]:
pattern = r"[a-yA-Y]*z[a-yA-Y]*"
string = "Brazil"
if match := re.search(pattern, string):
    print(match)
else:
    print("pattern not found")

<re.Match object; span=(0, 6), match='Brazil'>


##### 5. Write a Python program to match digits of length between 10 and 13.

In [ ]:
pattern = r"\d{10,13}"
string = "123123123123"
if match := re.search(pattern, string):
    print(match)
else:
    print("pattern not found")

<re.Match object; span=(0, 12), match='123123123123'>


##### 6. Write a Python program so that the full email addresses are extracted. (example: “For more information please write to the following email address: admissions@footballclub.com, Mrs Stepanie Smith will be delighted to assist you.”)

In [ ]:
pattern = r"\b[\w.-]+@[a-zA-Z.-]+(?<!\.)+\.[a-zA-Z]{2,3}\b"
string = "For more information please write to the following email address: admissions@footballclub.com, Mrs Stepanie Smith will be delighted to assist you."
if match := re.search(pattern, string):
    print(match)
else:
    print("pattern not found")

<re.Match object; span=(66, 93), match='admissions@footballclub.com'>


##### 7. E-mail Validation  
a) Write an email validating function, returning True or False depending on whether an email address is correctly formatted (e.g. john.smith@yahoo.com works, not  john.smith@yahoο, or  johnsmith@yahoo.com). No input should lead to an exception.
- Bonus. Extend your email validation in the first part so  
    b) it will accept not only  “.” but also  "-" and "_"  .  
    c) requires at least four characters after the ./-/_ symbol.  
    d) extensions with minimum 2 characters and maximum 4 characters shall be acceptable

- a) Write an email validating function, returning True or False depending on whether an email address is correctly formatted (e.g. john.smith@yahoo.com works, not john.smith@yahoο, or johnsmith@yahoo.com). No input should lead to an exception.

In [8]:
def email_validator(email: str) -> bool:
    return bool(re.search(pattern, email))

In [ ]:
pattern = "[\w.-]+@[a-zA-Z.-]+(?<!\.)\.[a-zA-Z]{2,3}"
sample_strings = [
    "john.smith@ya-hoo.com",
    "john.smith@yahoο",
    "johnsmith@yahoo..com",
]
list(zip(sample_strings, map(email_validator, sample_strings)))

[('john.smith@ya-hoo.com', True),
 ('john.smith@yahoο', False),
 ('johnsmith@yahoo..com', False)]

- b) it will accept not only “.” but also "-" and "" .

In [ ]:
# after experimenting '-' in brackets is auto-escaped if place at the end or start
pattern = "[\w.-]+@[a-zA-Z.-]+(?<![.-])[-._][a-zA-Z]{2,3}$"
sample_strings = [
    "john.smith@yahoo_com",
    "john.smith@yahoo.com",
    "john.smith@yahoo-com",
    "john.smith@yahoo@com",
]
list(zip(sample_strings, map(email_validator, sample_strings)))

[('john.smith@yahoo_com', True),
 ('john.smith@yahoo.com', True),
 ('john.smith@yahoo-com', True),
 ('john.smith@yahoo@com', False)]

- c) requires at least four characters after the ./-/ symbol.

In [ ]:
pattern = "[\w.-]+@[a-zA-Z.-]+(?<![.-])[-._][a-zA-Z]{4,}$"
sample_strings = [
    "john.smith@yahoo_comasd",
    "john.smith@yahoo.coms",
    "john.smith@yahoo-com",
    "john.smith@yahoo@com",
]
list(zip(sample_strings, map(email_validator, sample_strings)))

[('john.smith@yahoo_comasd', True),
 ('john.smith@yahoo.coms', True),
 ('john.smith@yahoo-com', False),
 ('john.smith@yahoo@com', False)]

- d) extensions with minimum 2 characters and maximum 4 characters shall be acceptable

In [ ]:
pattern = "[\w.-]+@[a-zA-Z.-]+(?<![.-])[-._][a-zA-Z]{2,4}$"
sample_strings = [
    "john.smith@yahoo_comasd",
    "john.smith@yahoo.coms",
    "john.smith@yahoo-com",
    "john.smith@yahoo@co",
]
list(zip(sample_strings, map(email_validator, sample_strings)))

[('john.smith@yahoo_comasd', False),
 ('john.smith@yahoo.coms', True),
 ('john.smith@yahoo-com', True),
 ('john.smith@yahoo@co', False)]

## Solo Practice

#### 1. Validate Email Address
Specification:
Match an email that:

Begins with one or more alphanumeric characters (including underscores or dots, but not consecutive dots and not starting or ending with them).

Contains exactly one “@” symbol.

The domain must be one or more alphanumeric characters (dots allowed but not consecutive), ending with a dot and a 2-4 letter TLD (only letters).

In [ ]:
match_emails = [
    "john.doe@example.com",
    "mary_jane123@sub.domain.org",
    "user.name@domain.co",
]
no_match_emails = [
    "john..doe@example.com",  # consecutive dots in local part
    "jane@domain",  # missing TLD
    "user@domain.toolong",  # TLD more than 4 letters
    "user@@domain.com",  # double @
]

In [ ]:
local_pattern = r"^\w(?:\w+\.)*\w*"
at_pattern = r"(?<!\.)@"
domain_pattern = r"\w(?:\w+\.)*\w*"
tld_pattern = r"\.[a-zA-Z]{2,4}$"
pattern = local_pattern + at_pattern + domain_pattern + tld_pattern

In [11]:
def pattern_validator(string: str, ignorecase=False) -> bool:
    if not ignorecase:
        return bool(re.search(pattern, string))
    else:
        return bool(re.search(pattern, string, re.IGNORECASE))

In [16]:
list(zip(match_emails, map(pattern_validator, match_emails)))

[('john.doe@example.com', True),
 ('mary_jane123@sub.domain.org', True),
 ('user.name@domain.co', True)]

In [17]:
list(zip(no_match_emails, map(pattern_validator, no_match_emails)))

[('john..doe@example.com', False),
 ('jane@domain', False),
 ('user@domain.toolong', False),
 ('user@@domain.com', False)]

In [ ]:
import rstr

attempt = rstr.xeger(pattern)
pattern = r"(?!.{50,})(?=.{0,20}$)(?=.{,20}$)(?=.{,20})?=.{,20}\\w(?:\\w+\\.)*\\w*(?<!\\.)@\\w(?:\\w+\\.)*\\w*\\.[a-zA-Z]{2,4}$"
rstr.xeger(pattern)

ModuleNotFoundError: No module named 'rstr'

#### 2. US Phone Number (Various Formats)
Specification:
Match a US phone number in one of these formats:

123-456-7890

(123) 456 7890

123 456 7890
Rules:

Area code can optionally be enclosed in parentheses.

Separators can be either a hyphen or a space.

In [ ]:
match_phones = [
    "123-456-7890",
    "(123) 456 7890",
    "(123)-456-7890",
    "123 456 7890",
]
no_match_phones = [
    "(123-456 7890",
    "(123 456-7890",
    "1234567890",  # no separators
    "(123-456-7890",  # mismatched parentheses
    "123.456.7890",  # dots not allowed
]

In [ ]:
pattern = r"^(?:\(\d{3}\)|\d{3})([ -])\d{3}\1\d{4}$"

In [ ]:
import rstr

rstr.xeger(pattern)

"+CUY>v>UC7huD TR,A A\\!\\v!+MU,V,leP hL2?qq/~0R=yU 'c56DRsFo7tY2moA32Uo8BgOgDajQy6hlZv58trKBef9KlSy5nK_dyokSwqMtbGOLkaDHBtGtlj0OjVmUsKoMtuWVkcg.o9pIHyMzHQCb27Ioc6MBX1TQVkNUZk_CaAnBuwMz8jxWCV0__NyDmoffTgI42HNsMpa1h.V4Kef4on_KULTX7PvNs.KtX7XwN0lcz4mv8K7MzGxy0atXqLUy0xfuDNZ2a8yCR_4O5UPvp3XrO3TeNSDw92hesgPYbahL5iu.WKm_E8Jv.Y_CygsnMPmD9I78aphHPRymQ6UM.PO5FCs_MbfKbY4ZqQ59me3e26WFeRC4vr17YTR2H8BWCzA3VoRfVqfauKMzzOQ5pMA8z2xCxACfsw.O2DoIZDvSJPiRYntXjiWw3PbbETDQIyqa5GGj3n7M3xGO9uEFXDyxSof_vifRDBenQ0s0J6fyOYKBcflzaiQDkgct.J1Bn4DOwBLO1O5wlRaVDHKkTZXIFYNTGOdf6di2w3t6OxvsxFRvaYzoz7MR9pa.IEuO.oFNQjC871fJ_a6vdXTit72MsfgHVHYm3Qp9HeXlAMePLVGPTiCP04WZrmKyWFNRLb_RdiRjA2JCjfCZcdEV3bHKZhwR.o7SHHVjw1QH.iGrA2q.1vvdngU6m3aAAaww5KB5h4424q8ClXKd8620BMoweXJlZR5.mhDsjKR4AathP8HMKFhlxPFNiwp56bKg3uUkOnV.HujSAPphYS5wvS2ggiN2k1xqFuqm1XdpV4wBhy7_rc_3MkOa2Hawo9p8ukFe8E2KZQt.jQFgsHkpELILldgvZ3Z1L3v2efzteug863NGlW0qKELNdE6elyFxHhd.VVPMmEtWhgX0kcs7Tq2HEM40Gza3aAN5BPjUWFZ1nuNPV2F6N0ta1agOT5iMhQ_dLANFa1RLGT9yzMT5eoWae6f0YfO

In [19]:
list(zip(match_phones, map(pattern_validator, match_phones)))

NameError: name 'match_phones' is not defined

In [20]:
list(zip(no_match_phones, map(pattern_validator, no_match_phones)))

NameError: name 'no_match_phones' is not defined

#### 3. Validate URL with HTTP/HTTPS
Specification:
Match a URL that:

Starts with http:// or https://.

Has a valid domain name: letters, digits, hyphens allowed.

Optionally may include a port (e.g., :8080).

Optionally ends with a path (starting with a slash) which may include alphanumerics and common URL symbols.

In [ ]:
match_urls = [
    "http://example.com",
    "https://sub-domain.example.co.uk:8080/path/to/resource",
    "http://example123.net/",
]
no_match_urls = [
    "https://sub-domain.example-.co.uk:8080/path/to/resource"  # dash next to dot
    "ftp://example.com",  # wrong protocol
    "http//example.com",  # missing colon
    "https://example",  # missing TLD
    "http://.com",  # missing domain name
]

In [ ]:
http_part = r"^https?://"
domain_part = r"(?:[a-zA-Z\d-]+(?<!-)\.)+"
tld_part = r"[a-zA-Z]{2,3}"
port_part = r"(?::\d{4})?"
path_part = r"(?:/\w+)*/?$"
pattern = http_part + domain_part + tld_part + port_part + path_part

In [23]:
list(zip(match_urls, map(pattern_validator, match_urls)))

[('http://example.com', True),
 ('https://sub-domain.example.co.uk:8080/path/to/resource', True),
 ('http://example123.net/', True)]

In [24]:
list(zip(no_match_urls, map(pattern_validator, no_match_urls)))

[('https://sub-domain.example-.co.uk:8080/path/to/resourceftp://example.com',
  False),
 ('http//example.com', False),
 ('https://example', False),
 ('http://.com', False)]

#### 4. Validate IPv4 Address
Specification:
Match a valid IPv4 address that:

Consists of four groups separated by a dot.

Each group is an integer from 0 to 255 (no leading zeros, except for 0 itself).

In [ ]:
match_ips = ["192.168.1.1", "10.0.0.255", "172.16.0.0"]
no_match_ips = [
    "256.100.50.0",  # 256 is out of range
    "192.168.01.1",  # leading zero in 01
    "192.168.1",  # only three groups
]

In [ ]:
triplet_group = r"(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]\d|\d)"
pattern = rf"^(?:{triplet_group}\.){{3}}{triplet_group}$"

In [27]:
list(zip(match_ips, map(pattern_validator, match_ips)))

[('192.168.1.1', True), ('10.0.0.255', True), ('172.16.0.0', True)]

In [28]:
list(zip(no_match_ips, map(pattern_validator, no_match_ips)))

[('256.100.50.0', False), ('192.168.01.1', False), ('192.168.1', False)]

#### 5. Extract Date in DD/MM/YYYY Format
Specification:
Match dates in the format DD/MM/YYYY where:

DD is 01-31,

MM is 01-12,

YYYY is any four-digit number.

Note: You don’t need full validation on days per month; assume any day 01–31 is allowed.

In [ ]:
match_dates = ["05/09/2020", "31/12/1999", "01/01/2021"]
no_match_dates = [
    "00/09/2020",  # day 00 invalid
    "32/01/2020",  # day 32 invalid
    "15-08-2020",  # wrong separator
]

In [ ]:
day = r"(?:3[01]|2\d|1\d|0[1-9])"
month = r"(?:0[1-9]|1[0-2])"
year = r"\d{4}"
pattern = rf"^{day}/{month}/{year}$"

In [31]:
list(zip(match_dates, map(pattern_validator, match_dates)))

[('05/09/2020', True), ('31/12/1999', True), ('01/01/2021', True)]

In [32]:
list(zip(no_match_dates, map(pattern_validator, no_match_dates)))

[('00/09/2020', False), ('32/01/2020', False), ('15-08-2020', False)]

#### 6. Validate Time Format (24-Hour)
Specification:
Match times in HH:MM format where:

HH is from 00 to 23.

MM is from 00 to 59.

In [ ]:
hour = r"(?:2[0-3]|[01]\d)"
minute = r"[0-5]\d"
pattern = rf"{hour}:{minute}"

In [ ]:
match_times = ["00:00", "12:34", "23:59"]
no_match_times = [
    "24:00",  # 24 is out of range
    "12:60",  # 60 is out of range
    "1:23",  # not exactly two digits for hour
]

In [35]:
list(zip(match_times, map(pattern_validator, match_times)))

[('00:00', True), ('12:34', True), ('23:59', True)]

In [36]:
list(zip(no_match_times, map(pattern_validator, no_match_times)))

[('24:00', False), ('12:60', False), ('1:23', False)]

#### 7. Validate Hexadecimal Number
Specification:
Match a hexadecimal number that:

Optionally starts with 0x or 0X.

Consists solely of hexadecimal digits (0–9, a–f, A–F) after the optional prefix.

Must have at least one digit.

In [ ]:
prefix = r"(?:0[xX])?"
value = r"[\da-fA-F]+"
pattern = rf"^{prefix}{value}$"

In [ ]:
match_hex = ["0x1A3F", "0Xabc", "3B7"]
no_match_hex = [
    "0x",  # no digits following prefix
    "0xGHI",  # invalid hex digits
    "x123",  # missing proper prefix and extra character
]

In [39]:
list(zip(match_hex, map(pattern_validator, match_hex)))

[('0x1A3F', True), ('0Xabc', True), ('3B7', True)]

In [40]:
list(zip(no_match_hex, map(pattern_validator, no_match_hex)))

[('0x', False), ('0xGHI', False), ('x123', False)]

#### 8. Validate Strong Password
Specification:
Match a password that must:

Be at least 8 characters long.

Contain at least one uppercase letter.

Contain at least one lowercase letter.

Contain at least one digit.

Contain at least one special character from !@#$%^&*()_+-=[]{}|;:,.<>?

In [ ]:
check_length = r"(?=.{8})"
check_upper = r"(?=.*[A-Z])"
check_lower = r"(?=.*[a-z])"
check_digit = r"(?=.*\d)"
check_special = r"(?=.*[!@#$%^&*()_+=\[\]{}|;:,.<>?-])"
pattern = rf"^{check_length}{check_upper}{check_lower}{check_digit}{check_special}.*$"

In [ ]:
match_passwords = ["Passw0rd!", "Str0ng@Pass", "A1b2C3$%"]
no_match_passwords = [
    "password",  # no uppercase, no digit, no special
    "PASSWORD123",  # no lowercase, no special
    "Passw0rd",  # missing special character
]

In [43]:
list(zip(match_passwords, map(pattern_validator, match_passwords)))

[('Passw0rd!', True), ('Str0ng@Pass', True), ('A1b2C3$%', True)]

In [ ]:
list(
    zip(no_match_passwords, map(pattern_validator, no_match_passwords))
)

[('password', False), ('PASSWORD123', False), ('Passw0rd', False)]

#### 9. Validate Credit Card Number (16 digits)
Specification:
Match a credit card number that:

Contains exactly 16 digits.

May optionally be divided into groups of 4 by spaces or hyphens (but not mixed).

In [ ]:
tetrad = r"[0-9]{4}"
tetrads_nr = 4
pattern = rf"^{tetrad}([- ]?)(?:{tetrad}\1){{{tetrads_nr - 2 if tetrads_nr > 1 else 0}}}{tetrad}$"

In [ ]:
match_cards = [
    "1234567812345678",
    "1234-5678-1234-5678",
    "1234 5678 1234 5678",
]
no_match_cards = [
    "1234-5678 1234-5678",  # mixed separators
    "12345678-12345678",  # wrong grouping
    "1234-5678-1234",  # too short
    "12345678901234567",  # too many digits
]

In [47]:
list(zip(match_cards, map(pattern_validator, match_cards)))

[('1234567812345678', True),
 ('1234-5678-1234-5678', True),
 ('1234 5678 1234 5678', True)]

In [48]:
list(zip(no_match_cards, map(pattern_validator, no_match_cards)))

[('1234-5678 1234-5678', False),
 ('12345678-12345678', False),
 ('1234-5678-1234', False),
 ('12345678901234567', False)]

#### 10. Validate US Social Security Number (SSN)
Specification:
Match an SSN that:

Follows the format XXX-XX-XXXX where X is a digit (0-9).

In [ ]:
pattern = r"^\d{3}-\d{2}-\d{4}$"

In [ ]:
match_ssn = ["123-45-6789", "000-12-3456"]
no_match_ssn = [
    "123456789",  # missing dashes
    "12-345-6789",  # wrong grouping
    "123-456-789",  # wrong number count in each group
]

In [51]:
list(zip(match_ssn, map(pattern_validator, match_ssn)))

[('123-45-6789', True), ('000-12-3456', True)]

In [52]:
list(zip(no_match_ssn, map(pattern_validator, no_match_ssn)))

[('123456789', False), ('12-345-6789', False), ('123-456-789', False)]

#### 11. Validate Hex Color Code
Specification:
Match a hex color code that:

Starts with a #.

Followed by exactly 6 hexadecimal digits (case insensitive).

In [ ]:
pattern = r"^#[\da-fA-F]{6}$"

In [ ]:
match_colors = ["#1A2B3C", "#abcdef", "#123456"]
no_match_colors = [
    "123456",  # missing hash
    "#123AB",  # only 5 hex digits
    "#1234567",  # 7 hex digits
    "#1G2B3C",  # invalid hex digit "G"
]

In [55]:
list(zip(match_colors, map(pattern_validator, match_colors)))

[('#1A2B3C', True), ('#abcdef', True), ('#123456', True)]

In [56]:
list(zip(no_match_colors, map(pattern_validator, no_match_colors)))

[('123456', False), ('#123AB', False), ('#1234567', False), ('#1G2B3C', False)]

#### 12. Validate ISBN-10 Format
Specification:
Match an ISBN-10 that:

Contains 10 characters.

The first 9 characters are digits.

The last character is either a digit or an uppercase X.

In [ ]:
pattern = r"^\d{9}[X\d]"

In [ ]:
match_isbn10 = ["0306406152", "123456789X"]
no_match_isbn10 = [
    "030640615",  # only 9 characters
    "123456789x",  # lowercase x not allowed (if the spec demands uppercase)
    "ABCDEFGHIJ",  # not digits
]

In [59]:
list(zip(match_isbn10, map(pattern_validator, match_isbn10)))

[('0306406152', True), ('123456789X', True)]

In [60]:
list(zip(no_match_isbn10, map(pattern_validator, no_match_isbn10)))

[('030640615', False), ('123456789x', False), ('ABCDEFGHIJ', False)]

#### 13. Validate US Zip Code
Specification:
Match a US zip code that:

Is either exactly 5 digits (e.g., 12345), or

5 digits followed by a hyphen and 4 more digits (e.g., 12345-6789).

In [ ]:
pattern = r"^\d{5}(?:-\d{4})?$"

In [ ]:
match_zipcodes = ["12345", "12345-6789"]
no_match_zipcodes = [
    "1234",  # too short
    "123456",  # too many digits without hyphen
    "12345_6789",  # wrong separator
]

In [63]:
list(zip(match_zipcodes, map(pattern_validator, match_zipcodes)))

[('12345', True), ('12345-6789', True)]

In [64]:
list(zip(no_match_zipcodes, map(pattern_validator, no_match_zipcodes)))

[('1234', False), ('123456', False), ('12345_6789', False)]

#### 14. Validate Simple HTML Opening Tag
Specification:
Match a simple HTML opening tag that:

Begins with < and ends with >.

Contains a tag name that starts with a letter and is followed by letters or digits.

Optionally, there may be one or more attributes. Each attribute is formatted as name="value" with no spaces within the attribute part.

In [ ]:
tag_name = r"[a-zA-Z][a-zA-Z\d]*"
attr_name = tag_name
attr = rf'{attr_name}=".*"(?=[> ])'
attrs = rf"(?:(?:{attr} )*{attr})"
pattern = rf"^<{tag_name}(?: {attrs})?>$"

In [ ]:
match_html = [
    "<div>",
    '<a href="link.html">',
    '<img src="image.png" alt="desc">',
]
no_match_html = [
    "<1div>",  # tag must start with a letter
    "<div class=example>",  # attribute value not in quotes
    "div>",  # missing starting <
]

In [67]:
list(zip(match_html, map(pattern_validator, match_html)))

[('<div>', True),
 ('<a href="link.html">', True),
 ('<img src="image.png" alt="desc">', True)]

In [68]:
list(zip(no_match_html, map(pattern_validator, no_match_html)))

[('<1div>', False), ('<div class=example>', False), ('div>', False)]

#### 15. Validate Time Range Format
Specification:
Match a time range string formatted as HH:MM-HH:MM where:

Both times follow the 24-hour HH:MM format (00:00 to 23:59).

There is a single hyphen separating the two valid times.

In [ ]:
hours = r"(?:2[0-3]|[01]\d)"
minutes = r"[0-5]\d"
time = rf"{hours}:{minutes}"
pattern = f"^{time}-{time}$"

In [ ]:
match_time_ranges = ["09:00-17:00", "00:00-23:59"]
no_match_time_ranges = [
    "9:00-17:00",  # first time not exactly HH:MM
    "09:00 to 17:00",  # wrong separator
    "24:00-17:00",  # invalid hour
]

In [71]:
list(zip(match_time_ranges, map(pattern_validator, match_time_ranges)))

[('09:00-17:00', True), ('00:00-23:59', True)]

In [ ]:
list(
    zip(
        no_match_time_ranges,
        map(pattern_validator, no_match_time_ranges),
    )
)

[('9:00-17:00', False), ('09:00 to 17:00', False), ('24:00-17:00', False)]

#### 16. Validate Full Name (First and Last with Capital Letters)
Specification:
Match a full name that:

Contains at least two words.

Each word starts with an uppercase letter (A–Z) followed by one or more lowercase letters.

Words are separated by a single space.

In [ ]:
name = r"[A-Z][a-z]+"
pattern = rf"^{name} {name}$"

In [ ]:
match_names = ["John Doe", "Alice Wonderland", "Mary Jane"]
no_match_names = [
    "john doe",  # lowercase initials
    "JohnDoe",  # no space
    "John doe",  # second word does not start with uppercase
    "J Doe",  # first name too short if you require more than one letter (depending on your exact rules)
]

In [75]:
list(zip(match_names, map(pattern_validator, match_names)))

[('John Doe', True), ('Alice Wonderland', True), ('Mary Jane', True)]

In [76]:
list(zip(no_match_names, map(pattern_validator, no_match_names)))

[('john doe', False),
 ('JohnDoe', False),
 ('John doe', False),
 ('J Doe', False)]

#### 17. Validate Unix File Path
Specification:
Match a Unix-like file path that:

Begins with a forward slash /.

Contains one or more directory names separated by /.

Directory names can be alphanumeric and may include underscores or hyphens.

Optionally ends with a file name with extension (e.g., .txt), following similar rules.

In [ ]:
valid_name = r"[\w-]+"
extension = r"(?:\.[a-zA-Z]{3})?"
pattern = rf"^(?:/{valid_name})+{extension}?$"

In [ ]:
match_paths = [
    "/usr/bin/python",
    "/home/user/Documents/report.txt",
    "/var/log/sys_log",
]
no_match_paths = [
    "usr/bin/python",  # missing starting slash
    "/home//user",  # double slash not allowed if you want one separator between names
    "/home/user/",  # ambiguous trailing slash if file name is required (depending on strictness)
]

In [79]:
list(zip(match_paths, map(pattern_validator, match_paths)))

[('/usr/bin/python', True),
 ('/home/user/Documents/report.txt', True),
 ('/var/log/sys_log', True)]

In [80]:
list(zip(no_match_ssn, map(pattern_validator, no_match_paths)))

[('123456789', False), ('12-345-6789', False), ('123-456-789', False)]

#### 18. Validate Python Variable Name
Specification:
Match a valid Python variable name that:

Starts with an underscore or letter (a–z, A–Z).

Followed by any combination of letters, digits, or underscores.

In [ ]:
pattern = r"^(?!\d)[\w\d]+$"

In [ ]:
match_variables = ["var_1", "_temp", "dataFrame2"]
no_match_variables = [
    "1stVar",  # starts with digit
    "var-name",  # hyphen not allowed
    "var!able",  # special character not allowed
]

In [83]:
list(zip(match_variables, map(pattern_validator, match_variables)))

[('var_1', True), ('_temp', True), ('dataFrame2', True)]

In [ ]:
list(
    zip(no_match_variables, map(pattern_validator, no_match_variables))
)

[('1stVar', False), ('var-name', False), ('var!able', False)]

#### 19. Match Alternating Vowels and Consonants (Simple Case)
Specification:
Match strings that:

Consist solely of alphabet characters.

The letters alternate strictly between vowels (a, e, i, o, u—case insensitive) and consonants.

For simplicity, assume the string starts with a vowel and has an even length.

In [ ]:
pattern = r"^([aeuoi](?=[a-z])[^aeuoi])+$"

In [ ]:
match_alternating = [
    "abab",  # a (vowel), b (consonant), a (vowel), b (consonant)
    "efoher",  # e (vowel), f (consonant), o (vowel), e (consonant)
]
no_match_alternating = [
    "aba",  # odd length
    "abba",  # letters do not alternate strictly
    "aeioua",  # multiple vowels in a row
]

In [ ]:
from functools import partial

pattern_validator_ic = partial(pattern_validator, ignorecase=True)

In [ ]:
list(
    zip(match_alternating, map(pattern_validator_ic, match_alternating))
)

[('abab', True), ('efoher', True)]

In [ ]:
list(
    zip(
        no_match_alternating,
        map(pattern_validator_ic, no_match_alternating),
    )
)

[('aba', False), ('abba', False), ('aeioua', False)]

#### 20. Validate Simple Mathematical Expression
Specification:
Match a mathematical expression that:

Consists of two integers (one or more digits) with an operator between them.

The operator can be one of +, -, *, or /.

Spaces around the operator are optional.

In [ ]:
pattern = r"^\d+ ?[-+/*] ?\d+$"

In [ ]:
match_math = ["3+4", "12 - 5", "100*20", "7 / 2"]
no_match_math = [
    "3 ++ 4",  # invalid operator sequence
    "12 -",  # missing second integer
    "+5 7",  # operator at the beginning
    "three+four",  # letters instead of digits
]

In [92]:
list(zip(match_math, map(pattern_validator, match_math)))

[('3+4', True), ('12 - 5', True), ('100*20', True), ('7 / 2', True)]

In [93]:
list(zip(no_match_math, map(pattern_validator, no_match_math)))

[('3 ++ 4', False), ('12 -', False), ('+5 7', False), ('three+four', False)]

##### 21. Validate a Suffix that Requires a Certain Prefix:
Specification:
Write a regex that matches the word “end” only if it is immediately preceded by a three-digit number.

In [ ]:
pattern = r"(?<=\b\d{3})end$"

In [ ]:
match_examples = ["123end", "999end"]
no_match_examples = [
    "12end",  # too short a prefix
    "1234end",  # more than three digits before
    "abc123end",  # non-digit characters before the digits
]

In [96]:
list(zip(match_examples, map(pattern_validator, match_examples)))

[('123end', True), ('999end', True)]

In [97]:
list(zip(no_match_examples, map(pattern_validator, no_match_examples)))

[('12end', False), ('1234end', False), ('abc123end', False)]

##### Negative Lookbehind for a Specific Pattern:
Specification:
Write a regex that matches the word "open" only if it is not preceded by "door".

In [ ]:
pattern = r"(?<!\bdoor )\bopen\b"

In [ ]:
match_examples = [
    "window open",
    "the open scene",
    "a door is closed, but room open",
]
no_match_examples = [
    "door open",  # should not match
    "big dooropen",  # when "door" is immediately adjacent to "open"
]

In [100]:
list(zip(match_examples, map(pattern_validator, match_examples)))

[('window open', True),
 ('the open scene', True),
 ('a door is closed, but room open', True)]

In [101]:
list(zip(no_match_examples, map(pattern_validator, no_match_examples)))

[('door open', False), ('big dooropen', False)]

##### 23. No Repeated Adjacent Words
Specification:
Create a regex that matches a sentence only if it does not contain any adjacent repeated words (case-insensitive). Use a negative lookahead with a backreference to detect repeats.
Example rule: "This is is a test." should fail because "is" appears twice consecutively; "This is a test." should pass.

In [ ]:
pattern = r"(?!.*\b(\w+)\b\W+\b\1)^.*$"

In [ ]:
match_no_repeat = [
    "This is a test.",
    "Hello world, hello again.",
    "One two three four.",
]
no_match_repeat = [
    "This is is a test.",
    "Well, well, what do we have here?",
    "Stop stop! Don't repeat.",
]

In [104]:
list(zip(match_no_repeat, map(pattern_validator_ic, match_no_repeat)))

[('This is a test.', True),
 ('Hello world, hello again.', True),
 ('One two three four.', True)]

In [105]:
list(zip(no_match_repeat, map(pattern_validator_ic, no_match_repeat)))

[('This is is a test.', False),
 ('Well, well, what do we have here?', False),
 ("Stop stop! Don't repeat.", False)]

##### 24. Validate a Date in YYYY-MM-DD with Leap Year Consideration (Partial)
Specification:
Write a regex to validate dates in the format YYYY-MM-DD with basic leap-year logic for February 29.

Year: any four digits (assume valid if divisible by 4 for simplicity, you need not validate century exceptions).

Month: 01 to 12.

Day: check up to 31 for months that allow it; for February, allow 29 only if the year is a leap year (divisible by 4), otherwise max 28.
Note: This regex won’t capture every real-life nuance but should use conditional (or nested) groups and lookahead/lookbehind if needed.

In [ ]:
lst = []
for i in range(0, 100, 4):
    lst.append(str(i).zfill(2))
lst[:3]
leap_years = "|".join(lst)
leap_years

'00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96'

In [ ]:
pattern = rf"^\d{{2}}(?:({leap_years})|\d{{2}})-(?:(02)|(04|06|09|11)|0[13578]|1[02])-(?(2)(?(1)(?:2\d|1\d|0[1-9])|(?:2[0-8]|1\d|0[1-9]))|(?:0[1-9]|1\d|2\d|3(?(3)0|[01])))$"

In [ ]:
match_dates = [
    "2020-02-29",  # 2020 is a leap year
    "2021-04-30",  # April has 30 days
    "2000-12-31",
]
no_match_dates = [
    "2021-02-29",  # 2021 is not a leap year
    "2020-04-31",  # April has only 30 days
    "1999-13-01",  # Invalid month
]

In [85]:
list(zip(match_dates, map(pattern_validator, match_dates)))

[('2020-02-29', True), ('2021-04-30', True), ('2000-12-31', True)]

In [86]:
list(zip(no_match_dates, map(pattern_validator, no_match_dates)))

[('2021-02-29', False), ('2020-04-31', False), ('1999-13-01', False)]

Let's try testing all the dates now!

In [ ]:
from datetime import datetime, date, timedelta


def date_generator():
    initial_date = date(2000, 1, 1)
    current_date = initial_date
    end_date = date(2100, 2, 28)

    while current_date < end_date:
        yield current_date
        current_date += timedelta(days=1)


date_list = map(
    lambda x: datetime.strftime(x, r"%Y-%m-%d"), date_generator()
)
all(map(pattern_validator, date_list))

True

The generator successfully passed all valid dates to our pattern_validator mapping and they all work.
If we had passed though 2100-02-29, it would pass too and this would not be correct.
Overall what a beautiful display of lazy iterator stacking though

##### 1. Validate Nested CSV Fields with Quotes and Escapes
Goal: Match valid CSV fields with:

Values either unquoted or "quoted"

Inside quoted values: double double-quotes ("") are allowed

Quoted values can include commas; unquoted cannot

In [ ]:
# value "quoted" "value, with, commas" "with ""escaped"" quotes"
quoted = r'(?:"(?:("?)[\w\s,]*\1)*")'
unquoted = r"(?:\w+$)"
pattern = rf"^{quoted}|{unquoted}"
pattern = r'^(?:(?:"(?:("?)[\w\s,]*\1)*")|\w+)$'

In [ ]:
match_values = [
    "value",
    '"quoted"',
    '"value, with, commas"',
    '"with ""escaped"" quotes"',
]
no_match_values = ['"unclosed', 'bad "quote"', "comma, inside"]

In [42]:
list(zip(match_values, map(pattern_validator, match_values)))

[('value', True),
 ('"quoted"', True),
 ('"value, with, commas"', True),
 ('"with ""escaped"" quotes"', True)]

In [43]:
list(zip(no_match_values, map(pattern_validator, no_match_values)))

[('"unclosed', False), ('bad "quote"', False), ('comma, inside', False)]

##### 2. Validate a Full URL with Optional Query Params
Goal: Match a URL with:

Protocol (http or https)

Optional subdomain

Domain

Optional port

Optional path

Optional query string

Use groups for each major component.

In [ ]:
pattern = (
    r"^(?P<protocol>https?://)(?P<subdomain>w{3}\.)?(?P<domain>(?:[\w-]+\.)+)(?P<tld>[a-zA-Z]{2,3})"
    r"(?P<port>:\d{4})?(?P<path>/\w+)*(?P<query>\?(?!&)(?:&?\w+=\w+)+)?$"
)

In [ ]:
match_values = [
    "http://example.com",
    "https://www.example.co.uk:8080",
    "https://www.example.co.uk:8080/path/to/page",
    "https://www.example.co.uk:8080/path/to/page?user=123&active=true",
    "https://www.example.co.asdadasdad.asdasdasd.uk:8080/path/to/page?user=123&active=true",
]
no_match_values = ["ftp://example.com", "http:/example.com", "http://"]

In [72]:
list(zip(match_values, map(pattern_validator, match_values)))

[('http://example.com', True),
 ('https://www.example.co.uk:8080', True),
 ('https://www.example.co.uk:8080/path/to/page', True),
 ('https://www.example.co.uk:8080/path/to/page?user=123&active=true', True),
 ('https://www.example.co.asdadasdad.asdasdasd.uk:8080/path/to/page?user=123&active=true',
  True)]

In [73]:
list(zip(no_match_values, map(pattern_validator, no_match_values)))

[('ftp://example.com', False),
 ('http:/example.com', False),
 ('http://', False)]

In [74]:
re.compile(pattern).search(match_values[4]).groupdict()

{'protocol': 'https://',
 'subdomain': 'www.',
 'domain': 'example.co.asdadasdad.asdasdasd.',
 'tld': 'uk',
 'port': ':8080',
 'path': '/page',
 'query': '?user=123&active=true'}

##### 3. Balanced Brackets Extractor
Goal: Match strings with balanced parentheses up to two levels deep using regex only (yes, this is pushing the limits).

Example: Use backreferences to simulate matching an opening ( with a closing ) without recursion.

In [ ]:
match_values = ["(a)", "(a(b)c)", "(x(y)z)"]
no_match_values = ["(a", "((a)", "a)", "a(b(c))"]

##### 4. Validate Version Strings With Optional Pre-release Tags
Goal: Match semantic versions like 1.0.0, 2.5.3-beta.1, 0.0.1-alpha, etc.

Major.Minor.Patch

Optional -alpha, -beta, -rc with optional dot-number

Use conditional group or backref for validating tag suffix

In [ ]:
match_values = ["1.0.0", "2.1.3-alpha", "3.2.1-beta.2", "1.2.3-rc.10"]
no_match_values = ["1.0", "v1.2.3", "1.0.0-unknown", "1.0.0-alpha-beta"]

##### 5. Validate Nested JSON-style Key-Value Pairs (Flat Only)
Goal: Match flat JSON-like key-value pairs using regex (no nesting):

`{"key1": "value1", "key2": "value2"}`
Keys and values in double quotes

Keys must be alphanumeric

Values can contain spaces, escaped quotes, but not unescaped "

Use lookaheads and backrefs to match structure.

In [ ]:
match_values = ['{"name": "John", "city": "New York"}']
no_match_values = [
    '{name: "John"}',
    '{"key": value}',
    '{"key": "bad"quote"}',
]

In [ ]:
text = """
The deadlines are: 2024-12-01, 2023-05-15, and 2025-01-30.
Please make sure to convert all dates to DD/MM/YYYY format.
"""
pattern = r"\b(?P<year>\d{4})-(?P<month>\d{2})-(?P<day>\d{2})\b"
print(re.sub(pattern, r"\g<day>/\g<month>/\g<year>", text))


The deadlines are: 01/12/2024, 15/05/2023, and 30/01/2025.
Please make sure to convert all dates to DD/MM/YYYY format.



In [ ]:
text = """
Here are some emails: alice@example.com, bob.smith@work.net, and someone123@domain.org.
These should all be redacted for privacy.
"""
pattern = r"\b\w+(?:\.\w+)*@\w+\.\w{2,3}\b"
print(re.sub(pattern, "[EMAIL]", text))


Here are some emails: [EMAIL], [EMAIL], and [EMAIL].
These should all be redacted for privacy.



In [ ]:
text = """
This is freaking awful. What the heck is going on?
No damn clue why everything is messed up.
"""


def replacer(match):
    return len(match.group()) * "*"


pattern = r"freaking|heck|damn"
print(re.sub(pattern, replacer, text))


This is ******** awful. What the **** is going on?
No **** clue why everything is messed up.



In [ ]:
text = """
Big shoutout to @johndoe, @jane_doe, and @admin for the update!
Follow them for more info.
"""
# @username with <a href="/user/username">@username</a>
pattern = r"@(\w+)\b"
replacement = r"<a href=\"/user/\1\">@\1</a>"
print(re.sub(pattern, replacement, text))


Big shoutout to <a href=\"/user/johndoe\">@johndoe</a>, <a href=\"/user/jane_doe\">@jane_doe</a>, and <a href=\"/user/admin\">@admin</a> for the update!
Follow them for more info.



In [ ]:
text = """
Wait... what?? Are you serious!!! This can't be real...
Oh no... not again!!!!
"""
pattern = r"([.?!])+"
print(re.sub(pattern, r"\1", text))


Wait. what? Are you serious! This can't be real.
Oh no. not again!



In [ ]:
text = """
Items: Book - $12.50, Pen - USD 3.00, Notebook - 1 dollar.
All prices must be normalized to this format: $XX.XX
"""


def replacer(match):
    if match.group(1):
        return "$" + match.group(1)
    else:
        return "$" + match.group(2) + ".00"


pattern = r"\b(?:USD (\d+\.\d{2}))|(\d+)* dollars?\b"
print(re.sub(pattern, replacer, text))


Items: Book - $12.50, Pen - $3.00, Notebook - $1.00.
All prices must be normalized to this format: $XX.XX



In [ ]:
text = """
Contacts: Alice - 123-456-7890, Bob - (987) 654-3210, Carl - 555.666.7777
We need to mask the middle 3 digits.
"""
pattern = r"(\b\d{3})\b(\W*)\d{3}\b(\W*)(\d{4})\b"


def replacer(match):
    return (
        match.group(1)
        + match.group(2)
        + "***"
        + match.group(3)
        + match.group(4)
    )


print(re.sub(pattern, replacer, text))


Contacts: Alice - 123-***-7890, Bob - (987) ***-3210, Carl - 555.***.7777
We need to mask the middle 3 digits.



Goal: Replace textual numbers (twenty-five, one hundred, etc.) with digits (25, 100, etc.).

Start by supporting a limited vocabulary: one to ninety-nine, and hundred.

Use a mapping dictionary and parse inside your function.

In [ ]:
text = """
He bought twenty-five apples and one hundred bananas.
They had seventy-two oranges and eleven pears.
"""

In [ ]:
numbers = {
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9,
    "ten": 10,
    "eleven": 11,
    "twelve": 12,
    "thirteen": 13,
    "hundred": 100,
    "thousand": 1000,
}

Goal: Replace all smart quotes (fancy ones like “”‘’) with regular ' and " using a mapping function.

In [ ]:
text = """
He said, “I’m not sure…” and then added, ‘but maybe’.
The symbols ‘ and “ need to be replaced with regular ' and ".
"""

Goal: Convert all DD Mon YYYY formats to YYYY-MM-DD.

So:

12 Jan 2024 → 2024-01-12

5 Feb 2023 → 2023-02-05

Use a function to:

Extract day, month abbreviation, and year

Convert the month name to number (Jan → 01, etc.)

Format as YYYY-MM-DD

In [ ]:
text = """
The event is on 12 Jan 2024. The deadline was 5 Feb 2023.
Next meeting is scheduled for 30 Mar 2025.
"""

##### Advanced Exercise: Selective Phone Masking Based on Area Code

Mask the middle 3 digits (like before), but only if the area code is one of:

123

555

987

Leave other numbers like 800-123-4567 unchanged.

💡 Challenge Details
Match multiple formats:

123-456-7890

(987) 654-3210

555.666.7777

Preserve separators.

Write a function to conditionally mask based on the captured area code.

Example Output:
Emergency contacts:
- Alice: 123-***-7890
- Bob: (987) ***-3210
- Carl: 555.***.7777
- Dave: 800-123-4567

In [ ]:
text = """
Emergency contacts:
- Alice: 123-456-7890
- Bob: (987) 654-3210
- Carl: 555.666.7777
- Dave: 800-123-4567

Only numbers with a local area code (starting with 123, 555, or 987) should be masked.
"""